internal/sops/awskms/keysource.go

```diff
@@ -0,0 +1,256 @@
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Encrypt takes a sops data key, encrypts it with KMS and stores the result in the EncryptedKey field
+func (key *MasterKey) Encrypt(dataKey []byte) error {
+	cfg, err := key.createKMSConfig()
+	if err != nil {
+		return err
+	}
+	client := kms.NewFromConfig(*cfg)
+	input := &kms.EncryptInput{
+		KeyId:     &key.Arn,
+		Plaintext: dataKey,
+	}
+	out, err := client.Encrypt(context.TODO(), input)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt sops data key with AWS KMS: %w", err)
+	}
+	key.EncryptedKey = base64.StdEncoding.EncodeToString(out.CiphertextBlob)
+	return nil
+}
+
+// EncryptIfNeeded encrypts the provided sops' data key and encrypts it if it hasn't been encrypted yet
+func (key *MasterKey) EncryptIfNeeded(dataKey []byte) error {
+	if key.EncryptedKey == "" {
+		return key.Encrypt(dataKey)
+	}
+	return nil
+}
+
+// Decrypt decrypts the EncryptedKey field with AWS KMS and returns the result.
+func (key *MasterKey) Decrypt() ([]byte, error) {
+	k, err := base64.StdEncoding.DecodeString(key.EncryptedKey)
+	if err != nil {
+		return nil, fmt.Errorf("error base64-decoding encrypted data key: %s", err)
+	}
+	cfg, err := key.createKMSConfig()
+	if err != nil {
+		return nil, err
+	}
+	client := kms.NewFromConfig(*cfg)
+	input := &kms.DecryptInput{
+		KeyId:             &key.Arn,
+		CiphertextBlob:    k,
+		EncryptionContext: key.EncryptionContext,
+	}
+	decrypted, err := client.Decrypt(context.TODO(), input)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt sops data key with AWS KMS: %w", err)
+	}
+	return decrypted.Plaintext, nil
+}
+
+// NeedsRotation returns whether the data key needs to be rotated or not.
+func (key *MasterKey) NeedsRotation() bool {
+	return time.Since(key.CreationDate) > (time.Hour * 24 * 30 * 6)
+}
+
+// ToString converts the key to a string representation
+func (key *MasterKey) ToString() string {
+	return key.Arn
+}
+
+// NewMasterKey creates a new MasterKey from an ARN, role and context, setting the creation date to the current date
+func NewMasterKey(arn string, role string, context map[string]string) *MasterKey {
+	return &MasterKey{
+		Arn:               arn,
+		Role:              role,
+		EncryptionContext: context,
+		CreationDate:      time.Now().UTC(),
+	}
+}
+
+// NewMasterKeyFromArn takes an ARN string and returns a new MasterKey for that ARN
+func NewMasterKeyFromArn(arn string, context map[string]string, awsProfile string) *MasterKey {
+	k := &MasterKey{}
+	arn = strings.Replace(arn, " ", "", -1)
+	roleIndex := strings.Index(arn, "+arn:aws:iam::")
+	if roleIndex > 0 {
+		k.Arn = arn[:roleIndex]
+		k.Role = arn[roleIndex+1:]
+	} else {
+		k.Arn = arn
+	}
+	k.EncryptionContext = context
+	k.CreationDate = time.Now().UTC()
+	return k
+}
+
+func (key MasterKey) createKMSConfig() (*aws.Config, error) {
+	cfg, err := config.LoadDefaultConfig(context.TODO(), func(lo *config.LoadOptions) error {
```

Comment:
Feel this could use a comment that if `key.credentialsProvider` is `nil`, the behavior of `LoadDefaultConfig` is to look at the environment.